### PR TITLE
Raise version min `PHP 7.3`.

### PR DESCRIPTION
**Pull request by terabytesoftw**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -82,11 +82,10 @@
         "paragonie/random_compat": ">=1"
     },
     "require-dev": {
-        "cweagans/composer-patches": "^1.7",
-        "phpunit/phpunit": "4.8.34",
         "cebe/indent": "~1.0.2",
-        "johnkary/phpunit-speedtrap": "^1.0",
         "dealerdirect/phpcodesniffer-composer-installer": "*",
+        "dms/phpunit-arraysubset-asserts": "^0.5",
+        "phpunit/phpunit": "9.6",
```
Comment (rob006) [inline composer.json:88]:
```suggestion
        "phpunit/phpunit": "^9.6",
```

**Review by bizley — APPROVED** [review]

**Review by samdark — CHANGES_REQUESTED** [review]

File: tests/data/validators/models/ValidatorTestTypedPropModel.php [hunk 2]
```diff
@@ -11,5 +11,8 @@
 
 class ValidatorTestTypedPropModel extends Model
 {
-    public array $arrayTypedProperty = [true, false];
+    /**
+     * @var array
+     */
+    public $arrayTypedProperty = [true, false];
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply tests/data/validators/models/ValidatorTestTypedPropModel.php]: done.

File: tests/IsOneOfAssert.php [hunk 3]
```diff
@@ -14,38 +14,34 @@
...
-
     /**
-     * IsOneOfAssert constructor.
-     * @param array $allowedValues
+     * @var array the expected values
      */
-    public function __construct(array $allowedValues)
+    private $allowedValues = [];
+
+    public function __construct($allowedValues)
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply tests/IsOneOfAssert.php]: done.

File: tests/IsOneOfAssert.php [hunk 4]
```diff
@@ -14,38 +14,34 @@
...
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    protected function matches($other)
+    protected function matches($other): bool
     {
-        return in_array($other, $this->allowedValues, false);
+        return in_array($other, $this->allowedValues);
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply tests/IsOneOfAssert.php]: done.

File: tests/framework/helpers/ArrayHelperTest.php [hunk 5]
```diff
@@ -136,14 +136,12 @@ public function testRemove()
...
 
     /**
+     * @requires PHP < 8.1
+     *
      * @return void
      */
     public function testRemoveWithFloat()
     {
-        if (version_compare(PHP_VERSION, '8.1.0', '>=')) {
-            $this->markTestSkipped('Using floats as array key is deprecated.');
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply tests/framework/helpers/ArrayHelperTest.php]: done.

File: tests/framework/helpers/ArrayHelperTest.php [hunk 6]
```diff
@@ -527,14 +525,12 @@ public function testMergeEmpty()
...
 
     /**
+     * @requires PHP < 8.1
+     *
      * @see https://github.com/yiisoft/yii2/pull/11549
      */
     public function testGetValueWithFloatKeys()
     {
-        if (version_compare(PHP_VERSION, '8.1.0', '>=')) {
-            $this->markTestSkipped('Using floats as array key is deprecated.');
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply tests/framework/helpers/ArrayHelperTest.php]: done.

File: tests/framework/helpers/ArrayHelperTest.php [hunk 7]
```diff
@@ -754,12 +750,11 @@ public function testKeyExists()
         $this->assertFalse(ArrayHelper::keyExists('c', $array, false));
     }
 
+    /**
+     * @requires PHP < 8.1
+     */
     public function testKeyExistsWithFloat()
     {
-        if (version_compare(PHP_VERSION, '8.1.0', '>=')) {
-            $this->markTestSkipped('Using floats as array key is deprecated.');
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply tests/framework/helpers/ArrayHelperTest.php:760]: done.

File: tests/framework/db/ActiveRecordTest.php [hunk 8]
```diff
@@ -1424,38 +1424,6 @@ public function testUnlinkAllViaTable()
         $this->assertEquals(5, $itemClass::find()->count());
     }
 
-    /**
-     * @requires PHP 5.6
-     */
-    public function testCastValues()
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply tests/framework/db/ActiveRecordTest.php:1430]: done.

File: tests/framework/di/ContainerTest.php [hunk 9]
```diff
@@ -480,36 +479,13 @@ public function testContainerSingletons()
         $this->assertSame($foo, $sameFoo);
     }
 
-    /**
-     * @requires PHP 5.6
-     */
-    public function testVariadicConstructor()
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply tests/framework/di/ContainerTest.php:486]: done.

File: tests/framework/di/ContainerTest.php [hunk 10]
```diff
@@ -480,36 +479,13 @@ public function testContainerSingletons()
...
-        }
-
-        $container = new Container();
-        $container->get('yiiunit\framework\di\stubs\Variadic');
-    }
-
-    /**
-     * @requires PHP 5.6
-     */
-    public function testVariadicCallable()
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply tests/framework/di/ContainerTest.php:499]: done.

**Comment by terabytesoftw** [comment]
image

@rob006 You still think 7.3 will be useful here.

**Comment by rob006** [comment]
@terabytesoftw What would be practical benefit of raising requirements to PHP 7.4? Yii 2.0 is in maintenance mode, we won't add new features and we can't refactor existing code using types because it would be a BC break (and probably a massive waste of time - such changes make more sense for Yii 2.2). So what features of PHP 7.4 would be so useful that it would be worth to cut \~5% of userbase?

**Comment by terabytesoftw** [comment]
I think we should then:

- Release of `2.0.49.4` with the current `master` and (remove the 8.4 tests) to leave the tests green, and leave the last update green, for `PHP 5.4 - 8.3`.
- merge php `7.3` in the master with `php 8.4 tests`.
- merge null implicit fixes.

With this we must have all the tests in green, review more PR, merge and release `2.0.50`.

@samdark @bizley @rob006 @mtangoo @schmunk42

**Comment by rob006** [comment]
I think we could skip `2.0.49.4`, we specially released `2.0.49.3` so all changes in master would go to 2.0.50.

**Comment by bizley** [comment]
Yes, let's go to 2.0.50 finally.

**Comment by terabytesoftw** [comment]
> Yes, let's go to 2.0.50 finally.

Ok then merge this PR, and the implicit null, to finish removing the PHP 8.4 deprecations.

**Comment by bizley** [comment]
@samdark please approve

**Comment by mtangoo** [comment]
Let's go :)

**Review by rhertogh — APPROVED** [review]

**Review by samdark — APPROVED** [review]

**Review by Arhell — APPROVED** [review]

**Comment by mtangoo** [comment]
Since we all approve this am merging it!
Thanks to @terabytesoftw for taking your time to prepare this, and everyone for making it reach thus far.
Cheers!

**Pull request merged**